### DISCUSSION TITLE
[question] Binary format of generated kernel binaries for CPUs

### DISCUSSION THREAD

**Discussion by jjfumero** [pocl/pocl#1470] [Q&A] [answered]
I am using [opencl-intercept-layer](https://github.com/intel/opencl-intercept-layer) to generate and visualise the generated assembly code for CPUs. 

I wonder the format of the binary generated by POCL. When I disassemble the binary emted by the `cliloader` tool, I get the following error. 

```bash
$ python disassemble_all_cpu.py /home/juan/CLIntercept_Dump/java 
Running "./disassemble_cpu.sh" to disassemble all isabin files in: /home/juan/CLIntercept_Dump/java:
Disassembling /home/juan/CLIntercept_Dump/java/CLI_0000_E82C00E5_0000_3C9D2991_CPU.bin
readelf: /home/juan/CLIntercept_Dump/java/CLI_0000_E82C00E5_0000_3C9D2991_CPU.bin: Error: Not an ELF file - it has the wrong magic bytes at the start
readelf: /home/juan/CLIntercept_Dump/java/CLI_0000_E82C00E5_0000_3C9D2991_CPU.bin: Error: Not an ELF file - it has the wrong magic bytes at the start
./disassemble_cpu.sh: line 24: printf: 0x: invalid hex number
Found 1 file(s) to disassemble.
```

I wonder if there is a better way to see the generated assembly code produced by POCL for CPUs. My interest is to analyze the vector instructions being generated from my compile-tool chain. Any pointers will be appreciated.

**Comment by pjaaskel** [comment, marked as answer]
PoCL's CPU driver binary format is a (currently) undocumented tar-like package of the kernel compiler cache directory structure. To investigate the kernel compilation results you can use the `POCL_CACHE_DIR` to change the directory where the kernel compiler stores its temps (the default is usually ~/.cache/pocl) and then you can make the compiler to leave all the intermediate results (parallel.bc is the final one before codegen) with `POCL_LEAVE_KERNEL_COMPILER_TEMP_FILES=1`.  You will see a temp directory structure with specialized parallel.bc for certain runtime properties (kernel name, local size, goffs=0, small grid) that is fed to codegen for producing the final CPU binary (also in that directory parallel.so).

All the envs are documented here: http://portablecl.org/docs/html/using.html#tuning-pocl-behavior-with-env-variables

**Reply by jjfumero** [reply]
Thanks @pjaaskel  for the suggestions. I did this and the temp files appear to be empty. 
This is all I got for a single kernel launch:

```bash
$ ll cache/
total 0
-rw-------. 1 juan juan 0 May 17 15:54 tempfile_l1CVoU
```

Probably I am missing something. As a side note, TornadoVM native code is not compiled with ``pkg-config --libs --cflags pocl`` but rather the generic flags for OpenCL (`-lOpenCL` and some optimization flags). I am not sure if this is related.

**Reply by pjaaskel** [reply]
Did you `export POCL_LEAVE_KERNEL_COMPILER_TEMP_FILES=1`? Also what might be unexpected is thanks to PoCL's aggressive kernel compiler specialization, we produce the final binary only at NDRange time when we know the WG dimensions, thus if you only clBuildProgram() you might not get the final results yet.

**Reply by jjfumero** [reply]
My mistake, Yes, I did  but I selected the wrong OpenCL device 😫. Now I can see the `parallel.bc` being generated. 
Thank you!
